// Contents: Nitrate package manifest format
// Author: Wesley Jones

syntax = "proto2";

package nitrate.no3.package;

message Package {
  message SemVersion {
    required uint32 major = 1;
    required uint32 minor = 2;
    required uint32 patch = 3;
  }

  message UUID {
    required uint64 hi = 1;
    required uint64 lo = 2;
  }

  enum Category {
    UNSPECIFIED = 0;
    APPLICATION = 1;
    LIBRARY = 2;
    STANDARD_LIBRARY = 3;
  }

  message Contact {
    enum Role {
      UNSPECIFIED = 0;
      OWNER = 1;
      CONTRIBUTOR = 2;
      MAINTAINER = 3;
      SUPPORT = 4;
    }

    required string name = 1;
    required string email = 2;
    optional string phone = 3;
    repeated Role roles = 4;
  }

  message Dependency {
    required UUID uuid = 1;
    required SemVersion version = 2;
  }

  message Platforms {
    repeated string allow = 1;
    repeated string deny = 2;
  }

  message OptimizationOptions {
    message SystemRequirements {
      required uint64 min_free_cores = 1;
      required uint64 min_free_memory = 2;
      required uint64 min_free_storage = 3;
    }

    message Switch {
      repeated string alpha = 1;
      repeated string beta = 2;
      repeated string gamma = 3;
      repeated string llvm = 4;
      repeated string lto = 5;
      repeated string runtime = 6;
    }

    message Rapid {
      required Switch switch = 1;
    }

    message Debug {
      required Switch switch = 1;
    }

    message Release {
      required Switch switch = 1;
    }

    required Rapid rapid = 1;
    required Debug debug = 2;
    required Release release = 3;
    required SystemRequirements requirements = 4;
  }

  message BlockchainLink {
    enum LinkCategory {
      UNSPECIFIED = 0;
      ECOSYSTEM_ROOT = 1;
      ECOSYSTEM_DOMAIN = 2;
      USER_ACCOUNT = 3;
      PACKAGE = 4;
      SUB_PACKAGE = 5;
    }

    message PublicKey {
      enum Type {
        UNSPECIFIED = 0;
        ED25519 = 1;
      }

      required Type type = 1;
      required bytes value = 2;
    }

    message Signature {
      enum Type {
        UNSPECIFIED = 0;
        ED25519 = 1;
      }

      required Type type = 1;
      required bytes value = 2;
    }

    required UUID uuid = 1;
    required LinkCategory category = 2;
    required PublicKey pubkey = 3;
    required Signature signature = 4;
  }

  required SemVersion format = 10;
  required string name = 20;
  required string description = 30;
  required string license = 40;
  required Category category = 50;
  required SemVersion version = 60;
  repeated Contact contacts = 80;
  required Platforms platforms = 90;
  repeated Dependency dependencies = 100;
  required OptimizationOptions optimization = 110;
  repeated BlockchainLink blockchain = 120;
}
